drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,328 @@
+		} else if (val->val1 == 10000) {
+			LOG_DBG("Configured range to %d", val->val1);
+			return mhz19b_send_config(dev, MHZ19B_CMD_SET_RANGE_10000);
+		} else {
+			return -ENOTSUP;
+		}
+
+	case SENSOR_ATTR_MHZ19B_ABC:
+		if (val->val1 == 0) {
+			LOG_DBG("%s ABC", "Disabled");
+			return mhz19b_send_config(dev, MHZ19B_CMD_SET_ABC_OFF);
+		} else if (val->val1 == 1) {
+			LOG_DBG("%s ABC", "Enabled");
+			return mhz19b_send_config(dev, MHZ19B_CMD_SET_ABC_ON);
+		} else {
+			return -ENOTSUP;
+		}
+
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int mhz19b_attr_get(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, struct sensor_value *val)
+{
+	struct mhz19b_data *data = dev->data;
+	int ret;
+
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		ret = mhz19b_poll_data(dev, MHZ19B_CMD_GET_RANGE);
+		val->val1 = (int32_t)data->data;
+		val->val2 = 0;
+		return ret;
+	case SENSOR_ATTR_MHZ19B_ABC:
+		ret = mhz19b_poll_data(dev, MHZ19B_CMD_GET_ABC);
+		val->val1 = (int32_t)data->data;
+		val->val2 = 0;
+		return ret;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int mhz19b_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+			      sensor_trigger_handler_t handler)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(trig);
+	ARG_UNUSED(handler);
+
+	return -ENOTSUP;
+}
+
+static int mhz19b_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE && chan != SENSOR_CHAN_POWER &&
+	    chan != SENSOR_CHAN_CURRENT) {
+		return -ENOTSUP;
+	}
+
+	return mhz19b_poll_data(dev, MHZ19B_CMD_GET_CO2);
+}
+
+static const struct sensor_driver_api mhz19b_api_funcs = {
+	.attr_set = mhz19b_attr_set,
+	.attr_get = mhz19b_attr_get,
+	.trigger_set = mhz19b_trigger_set,
+	.sample_fetch = mhz19b_sample_fetch,
+	.channel_get = mhz19b_channel_get,
+};
+
+static int mhz19b_init(const struct device *dev)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	struct sensor_value val;
+	int ret;
+
+	uart_irq_rx_disable(data->uart_dev);
+	uart_irq_tx_disable(data->uart_dev);
+
+	mhz19b_uart_flush(data->uart_dev);
+
+	uart_irq_callback_set(data->uart_dev, cfg->cb);
+
+	k_sem_init(&data->rx_sem, 0, 1);
+	k_sem_init(&data->tx_sem, 1, 1);
+
+	/* Configure default detection range */
+	val.val1 = cfg->range;
+
+	ret = mhz19b_attr_set(dev, SENSOR_CHAN_CO2, SENSOR_ATTR_FULL_SCALE, &val);
+	if (ret != 0) {
+		LOG_ERR("Error setting default range %d", val.val1);
+		return ret;
+	}
+
+	/* Configure ABC logic */
+	val.val1 = (uint32_t)cfg->abc_on;
+	ret = mhz19b_attr_set(dev, SENSOR_CHAN_CO2, SENSOR_ATTR_MHZ19B_ABC, &val);
+	if (ret != 0) {
+		LOG_ERR("Error setting default ABC %s", val.val1 == 1 ? "on" : "off");
+	}
+
+	return ret;
+}
+
+#define MHZ19B_UART_ISR_FN(inst)								\
+	static void mhz19b_uart_isr_##inst(const struct device *uart_dev, void *user_data)	\
```

Comment:
Do you really need a separate ISR function per instance? Can you use uart_irq_callback_user_data_set() instead of uart_irq_callback_set(), to save the sensor device in the user_data pointer?

Reply:
Done